### PR TITLE
NF: uses hash set/map which take size as init instead of capacity

### LINKED ISSUE #8337: Init capacity set capacity instead of size
This change sets the `initialCapacity` for `HashSet` and `HashMap`

My understanding is that these collections will expand their capacity when `capacity > intialCapcity * loadFactor` (maybe `>=`)

As `loadFactor < 1`, this means that by setting the capacity equal to the known collection size, we'll guarantee that each collection will go through a resize operation, which is what we want to avoid

_Originally posted by @david-allison-1 in https://github.com/ankidroid/Anki-Android/issues/7612#issuecomment-806495346_

### PR DESCRIPTION
**Pull request by Arthur-Milchior**
Fixes #8337

### CODE REVIEW

**Review by david-allison-1 — CHANGES_REQUESTED** [review]
```
  symbol:   class HashUtil
  location: package com.ichi2.utils
/home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/main/java/com/ichi2/themes/HtmlColors.java:3: error: cannot find symbol
import com.ichi2.utils.HashUtil;
```

Added a few places where the `HashUtil` is unused - this could indicate a problem, if it's fine, mark as resolved.

File: AnkiDroid/src/main/java/com/ichi2/libanki/Models.java [hunk 1]
```diff
@@ -1112,7 +1111,7 @@ public static boolean emptyStandardCard(JSONObject tmpl, Set<String> nonEmptyFie
     /**
      * Cache of getNamesOfFieldsContainingCloze
      * Computing hash of string is costly. However, hash is cashed in the string object, so this virtually ensure that
-     * given a card type, we don't need to recompute the hash.
+     * given a card type, we don't need to recompute the HashUtil.
```
[the remark this answers is not in the document]
Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/libanki/Models.java:1114]: I used script to do some replacement. But this one, I fail to see how it was catched. Make no sens

File: AnkiDroid/src/main/java/com/ichi2/libanki/template/ParsedNode.java [hunk 2]
```diff
@@ -1,15 +1,13 @@
 package com.ichi2.libanki.template;
 
 import android.content.Context;
-import android.text.TextUtils;
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/libanki/template/ParsedNode.java:4]: No need for this change

File: AnkiDroid/src/main/java/com/ichi2/libanki/AnkiPackageExporter.java [hunk 3]
```diff
@@ -22,6 +22,7 @@
 import com.ichi2.anki.CollectionHelper;
 import com.ichi2.anki.R;
 import com.ichi2.anki.exception.ImportExportException;
+import com.ichi2.utils.HashUtil;
```
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/libanki/AnkiPackageExporter.java:25]: remove unused header
```suggestion
```

**Review by mikehardy — CHANGES_REQUESTED** [review]
Seems fine to me - there is one unused import (doesn't codacy still send you email with new issues when it finds them? I got email on my last PR that had a codacy report showing I'd created issues and it pointed out 2 unused imports I forgot...)

Also needs a rebase / de-conflict

File: AnkiDroid/src/main/java/com/ichi2/libanki/AnkiPackageExporter.java (hunk 3 above)
Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/libanki/AnkiPackageExporter.java]: no, I never saw a codacy email

**Comment by mikehardy** [comment]
@Arthur-Milchior Something went terribly wrong in the rebase and now HashUtil, which is pretty important to this PR, is not in the change, causing predictable havoc...

**Comment by mikehardy** [comment]
Hey @Arthur-Milchior - can you patch this one up? Then we can get it in before moving on to the other PRs you are posting :)

File: AnkiDroid/src/main/java/com/ichi2/utils/HashUtil.java [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+     */
+    private static int capacity(int size) {
+        return Math.max((int) (size/.75f) + 1, 16);
+    }
+
+    public static <T> HashSet<T> HashSetInit(int size) {
+        return new HashSet<T>(capacity(size));
+    }
+    public static <T, U> HashMap<T, U> HashMapInit(int size) {
+        return new HashMap<T, U>(capacity(size));
```
[the remark this answers is not in the document]
Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/utils/HashUtil.java]: I spent some time to read about reflection but I don't understand how I could check that a function is not called. I don't see a way to change a method's code

**Comment by Arthur-Milchior** [comment]
Sorry, I missed the first message. Too many things right now

File: AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.java [hunk 5]
```diff
@@ -2034,7 +2034,7 @@ private void updateToolbar() {
 
     @NonNull
     private ArrayList<CustomToolbarButton> getToolbarButtons() {
-        Set<String> set = AnkiDroidApp.getSharedPrefs(this).getStringSet("note_editor_custom_buttons", new HashSet<>(0));
+        Set<String> set = AnkiDroidApp.getSharedPrefs(this).getStringSet("note_editor_custom_buttons", HashUtil.HashSetInit(0));
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.java:2037]: If we're getting into microoptimisations, we shouldn't be allocating the HashSet at all

Fine though

File: AnkiDroid/src/main/java/com/ichi2/utils/HashUtil.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (david-allison-1) [reply AnkiDroid/src/main/java/com/ichi2/utils/HashUtil.java:19]: You can't (easily) - but you can read the internal state of the object to see if threshold is changed

**Review by david-allison-1 — APPROVED** [review]
LGTM, had a play in the debugger and it seems good. A unit test would be very nice

**Comment by Arthur-Milchior** [comment]
I rebased and added what I thought would be correct test. But it tells that some field I see in the source code do not exists. So I assume it uses a different implementation. Don't see how to test here. Help wanted

**Comment by david-allison-1** [comment]
> I rebased and added what I thought would be correct test. But it tells that some field I see in the source code do not exists. So I assume it uses a different implementation. Don't see how to test here. Help wanted

(On my phone). It's probably JaCoCo. Try filtering on `!field.isSynthetic`

File: AnkiDroid/src/test/java/com/ichi2/utils/HashUtilTest.java [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+        return  thresholdField.getInt(map);
+    }
+
+    /** Threshold of a set. Not a setter*/
+    private int setThreshold(HashSet<?> set) throws NoSuchFieldException, IllegalAccessException {
+        return mapThreshold((HashMap) HashMap.class.getField("map").get(set));
+    }
+
+    @Test
+    public void NoReThreshold() throws NoSuchFieldException, IllegalAccessException {
```
Comment (david-allison-1) [inline AnkiDroid/src/test/java/com/ichi2/utils/HashUtilTest.java:24]:
```suggestion
    public void thresholdIsNotModifiedAfterCreationWithFixedSizeCollection() throws NoSuchFieldException, IllegalAccessException {
```

File: AnkiDroid/src/test/java/com/ichi2/utils/HashUtilTest.java [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+
+    /** Size of the table used to save a map.*/
+    private int mapThreshold(HashMap<?, ?> map) throws NoSuchFieldException, IllegalAccessException {
+        Field thresholdField = HashMap.class.getField("threshold");
+        return  thresholdField.getInt(map);
+    }
+
+    /** Threshold of a set. Not a setter*/
+    private int setThreshold(HashSet<?> set) throws NoSuchFieldException, IllegalAccessException {
+        return mapThreshold((HashMap) HashMap.class.getField("map").get(set));
```
Comment (david-allison-1) [inline AnkiDroid/src/test/java/com/ichi2/utils/HashUtilTest.java:20]:
```suggestion
        Field map = HashMap.class.getDeclaredField("map");
        map.setAccessible(true);
        return mapThreshold((HashMap) map.get(set));
```

File: AnkiDroid/src/test/java/com/ichi2/utils/HashUtilTest.java [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+import java.util.HashMap;
+import java.util.HashSet;
+import static org.junit.Assert.assertEquals;
+
+public class HashUtilTest {
+
+    /** Size of the table used to save a map.*/
+    private int mapThreshold(HashMap<?, ?> map) throws NoSuchFieldException, IllegalAccessException {
+        Field thresholdField = HashMap.class.getField("threshold");
+        return  thresholdField.getInt(map);
```
Comment (david-allison-1) [inline AnkiDroid/src/test/java/com/ichi2/utils/HashUtilTest.java:15]:
```suggestion
        Field thresholdField = HashMap.class.getDeclaredField("threshold");
        thresholdField.setAccessible(true);
        return thresholdField.getInt(map);
    }
```

**Review by david-allison-1 — COMMENTED** [review]
Try this

**Comment by david-allison-1** [comment]
Bad suggestion - apologies - split my code into 3 for ease of understanding

File: AnkiDroid/src/test/java/com/ichi2/utils/HashUtilTest.java [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+import static org.junit.Assert.assertEquals;
+
+public class HashUtilTest {
+
+    /** Size of the table used to save a map.*/
+    private int mapThreshold(HashMap<?, ?> map) throws NoSuchFieldException, IllegalAccessException {
+        Field thresholdField = HashMap.class.getDeclaredField("threshold");
+        thresholdField.setAccessible(true);
+        return thresholdField.getInt(map);
+    }
```
Comment (david-allison-1) [inline AnkiDroid/src/test/java/com/ichi2/utils/HashUtilTest.java:17]: Reminder to self: don't use suggestions for nontrivial fixes
```suggestion
```

**Comment by Arthur-Milchior** [comment]
May I just let it to you, will probably get easier.

**Comment by david-allison-1** [comment]
Sorry this took so long

Given the CI output, let's get this in without tests:

```
> Task :AnkiDroid:testDebugUnitTest
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by com.ichi2.utils.HashUtilTest (file:/home/runner/work/Anki-Android/Anki-Android/AnkiDroid/build/intermediates/javac/debugUnitTest/classes/) to field java.util.HashMap.threshold
WARNING: Please consider reporting this to the maintainers of com.ichi2.utils.HashUtilTest
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release

com.ichi2.utils.HashUtilTest > thresholdIsNotModifiedAfterCreationWithFixedSizeCollection FAILED
    java.lang.AssertionError: expected:<16> but was:<12>
        at org.junit.Assert.fail(Assert.java:89)
        at org.junit.Assert.failNotEquals(Assert.java:835)
        at org.junit.Assert.assertEquals(Assert.java:647)
        at org.junit.Assert.assertEquals(Assert.java:633)
        at com.ichi2.utils.HashUtilTest.thresholdIsNotModifiedAfterCreationWithFixedSizeCollection(HashUtilTest.java:37)
```

I'm happy it's working from experimentation

**Review by david-allison-1 — CHANGES_REQUESTED** [review]
adding this review flag: 

requested change: remove the tests given my comment above

**Comment by Arthur-Milchior** [comment]
Done

File: AnkiDroid/src/main/java/com/ichi2/utils/HashUtil.java [hunk 10]
```diff
@@ -0,0 +1,21 @@
+package com.ichi2.utils;
```
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/utils/HashUtil.java:1]: Missing copyright header check got you! :-)

```

/home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/main/java/com/ichi2/utils/HashUtil.java:1: Error: All files in AnkiDroid must contain a GPLv3-compatible copyright header [MissingCopyrightHeader from com.ichi2.anki.lint]
...
``

**Review by mikehardy — CHANGES_REQUESTED** [review]
I was all ready to hit the tempting + fun-to-hit merge button but we need a copyright header, then I think it's merge-able

**Comment by Arthur-Milchior** [comment]
done

**Review by david-allison-1 — APPROVED** [review]
LGTM, could do with a "Convert to Kotlin" for the new file, but that's not standardised yet

**Review by mikehardy — APPROVED** [review]
No need to keep messing around with this one - it took a conflict from a merge I just did but the code looks great on review now, I just went ahead and fixed the conflict if it passes CI I will squash-merge - thanks Arthur!

**Pull request merged**